sklearn/multioutput.py

```diff
@@ -359,3 +361,187 @@ def score(self, X, y):
                              format(n_outputs_, y.shape[1]))
         y_pred = self.predict(X)
         return np.mean(np.all(y == y_pred, axis=1))
+
+
+class ClassifierChain(BaseEstimator):
+    """A multi-label model that arranges binary classifiers into a chain. Each
+    model makes a prediction in the order specified by the chain using all
+    of the available features provided to the model plus the predictions of
+    models that are earlier in the chain.
+
+    Parameters
+    ----------
+    base_estimator : estimator
+        The base estimator from which the classifier chain is built.
+
+    random_state : int or RandomState, optional, default None
+        State or seed for random number generator.
+
+    order : list of integers, None, or 'random'
+        The order of the chain can be explicitly set by providing a list of
+        integers. For example, for a chain of length 5
+            order = [1, 3, 2, 4, 0]
+        means that the first model in the chain will make predictions for
+        column 1 in the Y matrix, the second model will make predictions
+        for column 3, etc.
+
+        If order is None the order will be determined by the
+        order of columns in the label matrix Y.
+            order = [0, 1, 2, ..., Y.shape[1] - 1]
+
+        If order is 'random' a random ordering will be used.
+
+    Attributes
+    ----------
+    classes_ : list of arrays of length len(estimators_) containing the
+        class labels for each estimator in the chain.
+
+    estimators_ : list
+        A list of copies of base_estimator. Once fit the estimators in
+        this list will be ordered as specified by the order attribute.
+        The original order of labels (as specified by Y) is recovered in
+        the predict method by indexing into the predictions with the
+        list of indices in the order attribute.
+
+
+    References
+    ----------
+    Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank, "Classifier
+    Chains for Multi-label Classification", 2009.
+
+    """
+
+    def __init__(self, base_estimator, random_state=None, order=None):
+        self.base_estimator = base_estimator
+        self.random_state = random_state
+        self.order = order
+
+    def fit(self, X, Y):
+        """Fit the model to data matrix X and targets Y.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Y : {array-like, sparse matrix}, shape (n_samples, n_classes)
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+
+        if sp.issparse(Y):
+            Y = Y.toarray()
+
+        if sp.issparse(X):
```

Comment:
I fixed the handling of sparse data and removed support for sparse labels (for now?) but I'm running into a problem where scipy raises an exception when you try to hstack an empty sparse array. This only seems to occur on some versions of python. Looking into possible solutions. Suggestions welcome.